### PR TITLE
Add Loongson MMI Support

### PR DESCRIPTION
**Pull request by zhuchen1911**
Hello!I do not want to close the request,but just want to sync with your updates.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -58,6 +58,29 @@ elseif(CMAKE_SYSTEM_PROCESSOR_LC MATCHES "arm*")
 elseif(CMAKE_SYSTEM_PROCESSOR_LC MATCHES "ppc*" OR
   CMAKE_SYSTEM_PROCESSOR_LC MATCHES "powerpc*")
   set(CPU_TYPE powerpc)
+elseif(CMAKE_SYSTEM_PROCESSOR_LC MATCHES "mips*")
```
Comment (mathstuf) [inline CMakeLists.txt]: Hmm. What are the processor names? `MATCHES` uses regexes, not globs, so this matches `mips`, `mip`, `mipss`, etc. Is `mips.*` intended?

The `ppc*` check above probably has the same thing.

**Comment by dcommander** [comment]
Sorry for the delay on this.  I am finally ready to integrate it, but I have a follow-up question:

Is there any way to identify a Loongson processor without running the assembler?  For instance, does it return a different value for `uname -p` (which is used to popular `CMAKE_SYSTEM_PROCESSOR`) than a regular MIPS CPU would return?

**Comment by dcommander** [comment]
To more formally state my case, there are two remaining things that I need before this code can be accepted into libjpeg-turbo:

1. A way to emulate a Loongson processor.  QEMU is acceptable, as long as it is available in pre-compiled binary form and there is an available Linux build, also in pre-compiled binary form, that supports Loongson.  MMI instructions should be supported in both.  As an example, here is the Docker image that I use for testing MIPS DSPr2 code: https://github.com/libjpeg-turbo/docker/tree/master/mips.  If there is no way for me to emulate Loongson MMI at the moment, then I cannot accept the code until such an emulator becomes available.  I must have the ability to independently verify the correctness of all code that is checked into libjpeg-turbo, and the ability to emulate the processor gives me additional assurances that the instruction set and architecture are mature and stable.  libjpeg-turbo, as a matter of policy, does not support architectures or instruction sets that are still in flux.
2. A way to differentiate Loongson from "regular" MIPS processors at build time without running the assembler.  The assembler should only be invoked when building the SIMD extensions.  Your specific implementation would break if building libjpeg-turbo for a "regular" MIPS processor and specifying `-DWITH_SIMD=0` to disable the SIMD extensions.  If Loongson cannot be differentiated at build time using normal methods (such as running `uname`), then it will be necessary to build libjpeg-turbo for Loongson using a CMake toolchain file that sets `CMAKE_SYSTEM_PROCESSOR` to a custom value (for example, "loongson") that our build system can detect.

Please comment this week.  If I do not hear from you by the end of the week, I will have no choice but to reject this PR for now, since libjpeg-turbo 1.6 beta needs to be released ASAP.

**Comment by dcommander** [comment]
@zhuchen1911 I don't think we are communicating.  You pushed a new commit, but that is not what I want.  I ultimately need to integrate this code myself, and in order to do that, I need answers to my questions above.

**Comment by zhuchen1911** [comment]
Hello! Here are our detailed responses about your two questions:
1.A way to emulate a Loongson processor. QEMU is acceptable, as long as it is available in pre-compiled binary form and there is an available Linux build, also in pre-compiled binary form, that supports Loongson. MMI instructions should be supported in both. As an example, here is the Docker image that I use for testing MIPS DSPr2 code: https://github.com/libjpeg-turbo/docker/tree/master/mips. If there is no way for me to emulate Loongson MMI at the moment, then I cannot accept the code until such an emulator becomes available. I must have the ability to independently verify the correctness of all code that is checked into libjpeg-turbo, and the ability to emulate the processor gives me additional assurances that the instruction set and architecture are mature and stable. libjpeg-turbo, as a matter of policy, does not support architectures or instruction sets that are still in flux.

A: We can provide you a remote docker environment for authentication, username, password as follows:

                 ssh root@cloud.loongnix.org -p 8023
                 Password: docker_jqnCLP
2.A way to differentiate Loongson from "regular" MIPS processors at build time without running the assembler. The assembler should only be invoked when building the SIMD extensions. Your specific implementation would break if building libjpeg-turbo for a "regular" MIPS processor and specifying -DWITH_SIMD=0 to disable the SIMD extensions. If Loongson cannot be differentiated at build time using normal methods (such as running uname), then it will be necessary to build libjpeg-turbo for Loongson using a CMake toolchain file that sets CMAKE_SYSTEM_PROCESSOR to a custom value (for example, "loongson") that our build system can detect.

A: I agree with your solution, because our cpu does not have a good automatic detection method, the use of the preparation toolchain.cmake specified "CMAKE_SYSTEM_PROCESSOR" way is a good solution.Specially，we should SET both the "CPU_TYPE" and "CMAKE_SYSTEM_PROCESSOR" to "loongson" because the "CPU_TYPE" is "NULL" in our system.

**Comment by dcommander** [comment]
Thank you very much.  I will proceed with integration.

**Comment by dcommander** [comment]
This has been integrated, with numerous code formatting and structure fixes.  In my testing, the MMI SIMD extensions accelerate RGB compression by ~70-101% (avg. 88%) and RGB decompression by 98-255% (avg. 167%.)

General critiques of your implementation:

* simd/jsimd.h should only contain prototypes for functions that are currently implemented.  Your implementation contained prototypes for all potential MMI functions.
* The jsimd_can\_\*() functions (in simd/loongson/jsimd.c) should return 0 and do nothing else if the corresponding algorithm is not implemented.
* In general, all intrinsics-based SIMD implementations should use a separate source file for every algorithm group (refer to the PowerPC Altivec implementation for an example.)  I would actually prefer that all SIMD implementations be organized like that, although the ARM and MIPS implementations currently are not.
* For now, I have worked around the numerous de-referenced type puns in your code, but I would strongly prefer it if you developed a more elegant solution for type casting between uint64_t and __m64 (a union, for instance.)
* As you can see, the Loongson section in simd/CMakeLists.txt should be similar to the PowerPC section, since both implementations use intrinsics.  Your implementation based this section upon the DSPr2 section, which was incorrect.  More specifically, your implementation unnecessarily enabled the ASM language in CMake.
* In general, please avoid borrowing code from the main libjpeg API library.  That code is under a slightly different license than the SIMD extensions, so pulling it into the SIMD modules creates difficulties for me as the maintainer.  As you can see, in places where it was necessary to borrow libjpeg code, I broke it out into a separate header file that retained the libjpeg license and attribution (jcsample.h, for instance.)  This is the difference between making something work and properly integrating it into a mature open source project.  Licenses matter.
* Your code formatting was a mess.  You can use my PowerPC Altivec extensions, or my modifications to your Loongson MMI extensions, as a guide for how to correctly format the code.  General guidelines:
  * Tabs should never be used.  The only exception is the TurboJPEG C API source files, which use tabs for historical reasons.
  * Wherever possible, code should not extend beyond Column 79.
  * 2-space indentation should be used.
  * Conditional blocks should keep the open bracket ({) on the same line as the if/else statement.
  * There should always be a space between "if" and "else" and the conditional expression.
  * There should always be a space between the conditional expression and the open bracket ({).
  * All arguments in a comma-separated argument list should be followed by a space.
  * There should never be a space between "*" and a pointer variable name.
  * Always use C-style comments (/* */), not C++-style comments (//), in C source files.
* In several places, you used very awkward while/if/else double conditionals when standard for() loops could achieve the same effect with fewer branches and more readable code.
* In many cases, you literally tried to copy the MMX code verbatim, but that's a poor approach to use for an intrinsics-based implementation, as it produces code that is hard to read.  The MMX implementation is legacy, so whereas it was reasonable to borrow some of its basic algorithms (given that MMI is also a 64-bit SIMD implementation), the code structure should have been borrowed more from the Altivec implementation.  One of the beauties of intrinsics is the ability to use three-operand instructions, which eliminates the need to perform as many register moves.  Also, since Loongson MMI can use up to 32 registers, there is no need to use the extreme loop unrolling and stack/memory swap techniques that were used by the MMX implementation (MMX only has eight registers.)  I was able to improve your performance by a few percent, simply by restructuring the code.  Additional performance can likely be obtained by streamlining the MMI implementation such that it takes advantage of the additional registers.
* libjpeg-turbo is a mature and stable project/product.  "TODO" notes should never be included in a pull request.

The integration of this feature was paid for by a Mozilla research grant, which has been exhausted.  I have no further funding to review additional enhancements to this feature, so in the future, I will reject Loongson patches that violate the above recommendations, unless your company is willing to pay for my labor to clean up and integrate the code.  It would take me less time to implement new MMI SIMD extensions myself rather than do the same cleanup that was necessary to integrate this code.  Hopefully you now have a better idea of what is expected from future code contributions.

For any further Loongson enhancements/bug fixes, please open a new pull request.  New algorithm implementations will likely miss 1.6, since I expect to freeze it within the next 1-2 weeks.

**Pull request closed without merging**